test/test_images/probe_helper/probe_helper_test.go

```diff
@@ -29,52 +49,346 @@ const (
+				}
+			} else if req.method == "DELETE" && req.url == testStorageGenerationRequest {
+				// This request indicates the client's intent to delete the object.
+				deletedEvent := newCloudEvent(
+					cloudEvent{
+						CeID:      "1234567890",
+						CeSubject: schemasv1.CloudStorageEventSubject("cloudstoragesource-probe-1234567890"),
+						CeType:    schemasv1.CloudStorageObjectDeletedEventType,
+						CeSource:  schemasv1.CloudStorageEventSource(testStorageBucket),
+					},
+				)
+				if res := c.Send(ctx, *deletedEvent); !cloudevents.IsACK(res) {
+					logger.Fatalf("Failed to send object deleted CloudEvent from the dummy CloudStorageSource: %v", res)
+				}
+			}
+		}
+	}
+}
+
+//
+type cloudEvent struct {
+	CeID      string
+	CeSubject string
+	CeSource  string
+	CeType    string
+}
+
+// Creates a new CloudEvent from raw data.
+func newCloudEvent(e cloudEvent) *cloudevents.Event {
 	event := cloudevents.NewEvent()
-	event.SetID("dummyID")
-	event.SetType("dummyType")
-	event.SetSource("dummySource")
-	if result := c.Send(context.Background(), event); !cloudevents.IsACK(result) {
-		t.Fatalf("failed to send cloudevent: %s", result.Error())
+	event.SetID(e.CeID)
+	event.SetSubject(e.CeSubject)
+	event.SetType(e.CeType)
+	event.SetSource(e.CeSource)
+	event.SetTime(time.Time{})
+	return &event
+}
+
+// Creates a new CloudEvent in the shape of probe events sent to the probe helper.
+func probeEvent(name, subject string) *cloudevents.Event {
+	return newCloudEvent(
+		cloudEvent{
+			CeID:      name + "-1234567890",
+			CeSubject: subject,
+			CeSource:  "probe-helper-test",
+			CeType:    name,
+		},
+	)
+}
+
+func newTestServer(handler func(w http.ResponseWriter, r *http.Request)) (*http.Client, func()) {
+	ts := httptest.NewTLSServer(http.HandlerFunc(handler))
+	tlsConf := &tls.Config{InsecureSkipVerify: true}
+	tr := &http.Transport{
+		TLSClientConfig: tlsConf,
+		DialTLS: func(netw, addr string) (net.Conn, error) {
+			return tls.Dial("tcp", ts.Listener.Addr().String(), tlsConf)
+		},
+	}
+	return &http.Client{Transport: tr}, func() {
+		tr.CloseIdleConnections()
+		ts.Close()
+	}
+}
+
+func testPubsubClient(ctx context.Context, t *testing.T, projectID string) (*pubsub.Client, func()) {
+	srv := pstest.NewServer()
+	conn, err := grpc.Dial(srv.Addr, grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("Failed to dial test pubsub connection: %v", err)
+	}
+	close := func() {
+		srv.Close()
+		conn.Close()
+	}
+	c, err := pubsub.NewClient(ctx, projectID, option.WithGRPCConn(conn))
+	if err != nil {
+		t.Fatalf("Failed to create test pubsub client: %v", err)
+	}
+	return c, close
+}
+
+type eventAndResult struct {
+	event      *cloudevents.Event
+	wantResult protocol.Result
+}
+
+type requestData struct {
+	method string
+	url    string
+	body   string
+}
+
+func TestProbeHelper(t *testing.T) {
+	t.Skip("Skip this test from running on Prow as it is only for local development.")
```

Comment:
Just to confirm, did you try running these tests and make sure they pass?

Context: We initially skipped the tests because currently the probeHelper uses a hard-coded port which may not be always available during tests. Ideally we should use httptest.NewServer

Reply:
I did run the tests and they work. I removed the line to skip the test in the latest commit, do you also mean I should replace the way the test Broker works (with the hardcoded port 9999) with httptest.NewServer?

Reply:
A hardcoded port isn't always guaranteed to be available during test and therefore creates flakiness. That's why we skipped the tests in CI/CD and only run it manually.

I am OK if we keep the Skip or we use something like httptest.NewServer to dynamically assign a port.